OpenTrons/opentrons-api
# feat(api): enforce only one item in any location

# Overview

Previously, we had to allow loading multiple items (hopefully of the same item, just multiple times) in any location to accommodate for LPC happening in the same run. Now with LPC happening in separate maintenance runs, we do not have to support this anymore. This PR removes all allowances of loading multiple items in a given location. Note, a 'location' means a very specific location like a particular slot or a module or a labware.

This is required in order to perform robust deck conflict checks without having to account for the impractical case of multiple items loaded in any given location (also makes the code cleaner).

# Test Plan

- Test protocols that load multiple items in the same location and verify that an error is raised
- Start an http run and try loading multiple items in the same location and verify that an error response is received
- Verify that labware stacking or loading labware on modules still completes successfully

# Changelog

- added checks in `loadLabware` and `loadModule` to verify that the specified load location is not occupied
- removed the labware & module ids parsing filter logic that was used to disregard equipment added to the engine during setup
- updated the existing JSON test protocols by removing the load trash commands in protocols. I tested the actual protocols exported by PD from 6.0.0 until the latest one and verified that none of the protocols _load_ the fixed trash explicitly. So unless someone manually edits a JSON protocol to add a load trash command, no protocols in the field should have this.
- updated the tests to reflect the updated the JSON protocols

# Review requests

- Check if I have missed any deck conflict check. I haven't added any checks for addressable areas but should I? They are a bit complicated as in they don't always reflect an equipment in a location..
- any tests I've missed?

# Risk assessment

- This PR adds a small risk of breaking protocols written outside our API & PD recommendations. Both PD & Python API have had checks for preventing loading of multiple equipment in one location, except for a bug that allowed loading multiple labware on a module. So unless a user has hand-edited JSON protocols or wrote python protocols that exploited the mentioned bug in their protocols, protocols in the field should not break. Also for protocols that did exploit this loophole, their runs had chances of getting into unexpected states so this change makes sure that doesn't happen from now on.
